### PR TITLE
feat: CCIP-2465 ccip-enumerableMap for addressToBytesMap

### PR DESCRIPTION
**Pull request by defistar**
# EnumerableMap for mapping Bytes32 to Bytes

## JIRA Ticket
https://smartcontract-it.atlassian.net/browse/CCIP-2465

## Background
- CCIP has rateLimiter with tokenMapping from local to remote chain
- The mappings are grouped by per remoteChainSelector to its localTokenAddress => remoteTokenAddress 
- at moment the mapping for tokenAddresses are using 
```
mapping(uint64 remoteChainSelector => EnumerableMapAddresses.AddressToBytes32Map tokensLocalToRemote) internal
    s_rateLimitedTokensLocalToRemote;
```

this needs to be changed to use 

```
mapping(uint64 remoteChainSelector => EnumerableMapAddresses.AddressToBytesMap tokensLocalToRemote) internal
    s_rateLimitedTokensLocalToRemote;
```

## Change List
1. new library `EnumerableMapBytes32.sol` which contains mapping from Bytes32 => Bytes
2. use the new library in `EnumerableMapAddresses.sol` for mapping arrangement functions for Address => Bytes

### CODE REVIEW

File: contracts/src/v0.8/shared/enumerable/EnumerableMapAddresses.sol [hunk 1]
```diff
@@ -2,17 +2,19 @@
 pragma solidity ^0.8.0;
 
 import {EnumerableMap} from "../../vendor/openzeppelin-solidity/v4.8.3/contracts/utils/structs/EnumerableMap.sol";
+import "./CCIPEnumerableMap.sol";
```
Comment (RensR) [inline contracts/src/v0.8/shared/enumerable/EnumerableMapAddresses.sol]: nit: use explicit imports

File: contracts/src/v0.8/shared/enumerable/CCIPEnumerableMap.sol [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+  }
+
+  /**
+   * @dev Returns the value associated with `key`. O(1).
+   *
+   * Requirements:
+   *
+   * - `key` must be in the map.
+   */
+  // solhint-disable-next-line chainlink-solidity/prefix-internal-functions-with-underscore
```
Comment (RensR) [inline contracts/src/v0.8/shared/enumerable/CCIPEnumerableMap.sol]: nit: we can turn off this rule for the entire file instead

File: contracts/src/v0.8/shared/enumerable/EnumerableMapAddresses.sol [hunk 3]
```diff
@@ -2,17 +2,19 @@
...
   using EnumerableMap for EnumerableMap.UintToAddressMap;
   using EnumerableMap for EnumerableMap.Bytes32ToBytes32Map;
+  using CCIPEnumerableMap for CCIPEnumerableMap.Bytes32ToBytesMap;
 
   struct AddressToAddressMap {
     EnumerableMap.UintToAddressMap _inner;
   }
-
   // solhint-disable-next-line chainlink-solidity/prefix-internal-functions-with-underscore
+
```
Comment (RensR) [inline contracts/src/v0.8/shared/enumerable/EnumerableMapAddresses.sol]: nit: whitespace

File: contracts/src/v0.8/shared/enumerable/CCIPEnumerableMap.sol [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+ * ====
+ * Trying to delete such a structure from storage will likely result in data corruption, rendering the structure
+ * unusable.
+ * See https://github.com/ethereum/solidity/pull/11843[ethereum/solidity#11843] for more info.
+ *
+ * In order to clean an CCIPEnumerableMap, you can either remove all elements one by one or create a fresh instance using an
+ * array of CCIPEnumerableMap.
+ * ====
+ */
+library CCIPEnumerableMap {
```
Comment (RensR) [inline contracts/src/v0.8/shared/enumerable/CCIPEnumerableMap.sol]: let's remove the CCIP naming

File: contracts/src/v0.8/shared/enumerable/EnumerableMapAddresses.sol [hunk 5]
```diff
@@ -137,4 +139,88 @@ library EnumerableMapAddresses {
   function get(AddressToBytes32Map storage map, address key) internal view returns (bytes32) {
     return map._inner.get(bytes32(uint256(uint160(key))));
   }
+
+  struct AddressToBytesMap {
```
Comment (RensR) [inline contracts/src/v0.8/shared/enumerable/EnumerableMapAddresses.sol:142]: Let's create a new file for this and not put it in `EnumerableMapAddresses.sol`

Comment (elatoskinas) [reply contracts/src/v0.8/shared/enumerable/EnumerableMapAddresses.sol:142]: Don't think we need a new file, this struct is logically grouped to contain all `AddressToX` maps - same structure as OpenZeppelin's lib

File: contracts/src/v0.8/shared/enumerable/CCIPEnumerableMap.sol (hunk 4 above)
Comment (elatoskinas) [reply contracts/src/v0.8/shared/enumerable/CCIPEnumerableMap.sol]: +1, something like `EnumerableMapBytes32` to be consistent with `EnumerableMapAddresses`

File: contracts/src/v0.8/shared/enumerable/CCIPEnumerableMap.sol [hunk 6]
```diff
@@ -0,0 +1,143 @@
...
+ * - `bytes32 -> bytes` (`Bytes32ToBytes`)
+ *
+ * [WARNING]
+ * ====
+ * Trying to delete such a structure from storage will likely result in data corruption, rendering the structure
+ * unusable.
+ * See https://github.com/ethereum/solidity/pull/11843[ethereum/solidity#11843] for more info.
+ *
+ * In order to clean an CCIPEnumerableMap, you can either remove all elements one by one or create a fresh instance using an
+ * array of CCIPEnumerableMap.
```
Comment (elatoskinas) [inline contracts/src/v0.8/shared/enumerable/CCIPEnumerableMap.sol]: suggestion: CCIPEnumerableMap is a library - we cannot create an array of these. Also, creating a new map will not clean the old map unless we delete the elements. We should reword this to something like:

> In order to clean up an CCIPEnumerableMap, you should remove all elements one by one.

File: contracts/src/v0.8/shared/enumerable/CCIPEnumerableMap.sol [hunk 7]
```diff
@@ -0,0 +1,143 @@
...
+ * ====
+ */
+library CCIPEnumerableMap {
+  using EnumerableSet for EnumerableSet.Bytes32Set;
+
+  // Define a custom error
```
Comment (elatoskinas) [inline contracts/src/v0.8/shared/enumerable/CCIPEnumerableMap.sol]: nit: redundant comment, we can remove

File: contracts/src/v0.8/shared/enumerable/CCIPEnumerableMap.sol [hunk 8]
```diff
@@ -0,0 +1,143 @@
...
+  error NonexistentKeyError();
+
+  // To implement this library for multiple types with as little code
```
Comment (elatoskinas) [inline contracts/src/v0.8/shared/enumerable/CCIPEnumerableMap.sol]: nit: this whole comment block does not apply (L47-55) - we do not have `bytes32` as values, and we do not have a generic map that we re-use for other maps - we just have one `Bytes32ToBytesMap`. We can remove

File: contracts/src/v0.8/shared/enumerable/CCIPEnumerableMap.sol [hunk 9]
```diff
@@ -0,0 +1,143 @@
...
+
+  /**
+   * @dev Removes a key-value pair from a map. O(1).
+   *
+   * Returns true if the key was removed from the map, that is if it was present.
+   */
+  // solhint-disable-next-line chainlink-solidity/prefix-internal-functions-with-underscore
+  function remove(Bytes32ToBytesMap storage map, bytes32 key) internal returns (bool) {
+    delete map._values[key];
+    return map._keys.remove(key);
```
Comment (elatoskinas) [inline contracts/src/v0.8/shared/enumerable/CCIPEnumerableMap.sol]: q: does this correctly remove all bytes values, since `bytes` is a dynamic array?

File: contracts/src/v0.8/shared/test/enumerable/EnumerableMapAddresses.t.sol [hunk 10]
```diff
@@ -119,11 +156,21 @@ contract EnumerableMapAddresses_at is EnumerableMapAddressesTest {
...
     assertTrue(value == MOCK_BYTES32_VALUE);
   }
+
+  function testBytesAtSuccess() public {
+    assertTrue(s_addressToBytesMap.length() == 0);
+    assertTrue(s_addressToBytesMap.set(address(this), MOCK_BYTES_VALUE));
+    assertTrue(s_addressToBytesMap.length() == 1);
+    (address key, bytes memory value) = s_addressToBytesMap.at(0);
+    assertTrue(key == address(this));
+    assertTrue(keccak256(value) == keccak256(MOCK_BYTES_VALUE));
```
Comment (elatoskinas) [inline contracts/src/v0.8/shared/test/enumerable/EnumerableMapAddresses.t.sol]: suggestion: foundry supports `assertEq`:

```suggestion
    assertEq(value, MOCK_BYTES_VALUE);
```

File: contracts/src/v0.8/shared/test/enumerable/EnumerableMapAddresses.t.sol [hunk 11]
```diff
@@ -142,11 +189,21 @@ contract EnumerableMapAddresses_tryGet is EnumerableMapAddressesTest {
...
     assertTrue(value == MOCK_BYTES32_VALUE);
   }
+
+  function testBytesTryGetSuccess() public {
+    assertTrue(!s_addressToBytesMap.contains(address(this)));
+    assertTrue(s_addressToBytesMap.set(address(this), MOCK_BYTES_VALUE));
+    assertTrue(s_addressToBytesMap.contains(address(this)));
+    (bool success, bytes memory value) = s_addressToBytesMap.tryGet(address(this));
+    assertTrue(success);
+    assertTrue(keccak256(value) == keccak256(MOCK_BYTES_VALUE));
```
Comment (elatoskinas) [inline contracts/src/v0.8/shared/test/enumerable/EnumerableMapAddresses.t.sol]:
```suggestion
    assertEq(value, MOCK_BYTES_VALUE);
```

File: contracts/src/v0.8/shared/test/enumerable/EnumerableMapAddresses.t.sol [hunk 12]
```diff
@@ -161,16 +218,34 @@ contract EnumerableMapAddresses_get is EnumerableMapAddressesTest {
     assertTrue(s_addressToBytes32Map.contains(address(this)));
     assertTrue(s_addressToBytes32Map.get(address(this)) == MOCK_BYTES32_VALUE);
   }
+
+  function testBytesGetSuccess() public {
+    assertTrue(!s_addressToBytesMap.contains(address(this)));
+    assertTrue(s_addressToBytesMap.set(address(this), MOCK_BYTES_VALUE));
+    assertTrue(s_addressToBytesMap.contains(address(this)));
+    assertTrue(keccak256(s_addressToBytesMap.get(address(this))) == keccak256(MOCK_BYTES_VALUE));
```
Comment (elatoskinas) [inline contracts/src/v0.8/shared/test/enumerable/EnumerableMapAddresses.t.sol]:
```suggestion
    assertEq(s_addressToBytesMap.get(address(this)), MOCK_BYTES_VALUE);
```

File: contracts/src/v0.8/shared/test/enumerable/EnumerableMapAddresses.t.sol [hunk 13]
```diff
@@ -161,16 +218,34 @@ contract EnumerableMapAddresses_get is EnumerableMapAddressesTest {
...
     assertTrue(s_addressToAddressMap.set(address(this), address(this)));
     assertTrue(s_addressToAddressMap.contains(address(this)));
     assertTrue(s_addressToAddressMap.get(address(this), "EnumerableMapAddresses: nonexistent key") == address(this));
   }
+
+  function testBytesGetErrorMessageSuccess() public {
+    assertTrue(!s_addressToBytesMap.contains(address(this)));
+    assertTrue(s_addressToBytesMap.set(address(this), MOCK_BYTES_VALUE));
+    assertTrue(s_addressToBytesMap.contains(address(this)));
+    assertTrue(
```
Comment (elatoskinas) [inline contracts/src/v0.8/shared/test/enumerable/EnumerableMapAddresses.t.sol]: suggestion: same here - you can use `assertEq`

File: contracts/src/v0.8/shared/enumerable/CCIPEnumerableMap.sol (hunk 9 above)
Comment (defistar) [reply contracts/src/v0.8/shared/enumerable/CCIPEnumerableMap.sol]: added a test for this. will test on varying bytes size

File: contracts/src/v0.8/shared/enumerable/EnumerableMapAddresses.sol (hunk 5 above)
Comment (defistar) [reply contracts/src/v0.8/shared/enumerable/EnumerableMapAddresses.sol:142]: yes logical gtouping of all Address based mappings

File: contracts/src/v0.8/shared/test/enumerable/EnumerableMapAddresses.t.sol (hunk 13 above)
Comment (defistar) [reply contracts/src/v0.8/shared/test/enumerable/EnumerableMapAddresses.t.sol]: updated

File: contracts/src/v0.8/shared/test/enumerable/EnumerableMapAddresses.t.sol (hunk 12 above)
Comment (defistar) [reply contracts/src/v0.8/shared/test/enumerable/EnumerableMapAddresses.t.sol]: updated

File: contracts/src/v0.8/shared/test/enumerable/EnumerableMapAddresses.t.sol (hunk 11 above)
Comment (defistar) [reply contracts/src/v0.8/shared/test/enumerable/EnumerableMapAddresses.t.sol]: updated

File: contracts/src/v0.8/shared/enumerable/CCIPEnumerableMap.sol (hunk 8 above)
Comment (defistar) [reply contracts/src/v0.8/shared/enumerable/CCIPEnumerableMap.sol]: applied

File: contracts/src/v0.8/shared/enumerable/CCIPEnumerableMap.sol (hunk 7 above)
Comment (defistar) [reply contracts/src/v0.8/shared/enumerable/CCIPEnumerableMap.sol]: removed

File: contracts/src/v0.8/shared/enumerable/CCIPEnumerableMap.sol (hunk 6 above)
Comment (defistar) [reply contracts/src/v0.8/shared/enumerable/CCIPEnumerableMap.sol]: update

File: contracts/.changeset/nice-bees-do.md [hunk 14]
```diff
@@ -0,0 +1,5 @@
+---
+'@chainlink/contracts': minor
+---
+
+EnumerableMap Library needed for mapping Address as Bytes32 Key and Value as Bytes
```
Comment (elatoskinas) [inline contracts/.changeset/nice-bees-do.md]: nit: wording:
```suggestion
EnumerableMap Library for an Address to Bytes mapping
```

File: contracts/src/v0.8/shared/test/enumerable/EnumerableMapAddresses.t.sol [hunk 15]
```diff
@@ -75,63 +97,89 @@ contract EnumerableMapAddresses_contains is EnumerableMapAddressesTest {
...
+    assertEq(value, MOCK_BYTES32_VALUE);
+  }
+
+  function testBytesAtSuccess() public {
+    assertEq(s_addressToBytesMap.length(), 0);
+    assertTrue(s_addressToBytesMap.set(address(this), MOCK_BYTES_VALUE));
+    assertEq(s_addressToBytesMap.length(), 1);
+    (address key, bytes memory value) = s_addressToBytesMap.at(0);
+    assertEq(key, address(this));
+    assertEq(keccak256(value), keccak256(MOCK_BYTES_VALUE));
```
Comment (elatoskinas) [inline contracts/src/v0.8/shared/test/enumerable/EnumerableMapAddresses.t.sol]: suggestion: no need for `keccak`, you can compare directly. Foundry implements this function:

> function assertEq(bytes memory left, bytes memory right) internal;

```suggestion
    assertEq(value,MOCK_BYTES_VALUE);
```

File: contracts/src/v0.8/shared/test/enumerable/EnumerableMapAddresses.t.sol [hunk 16]
```diff
@@ -140,37 +188,62 @@ contract EnumerableMapAddresses_tryGet is EnumerableMapAddressesTest {
...
+    assertEq(value, MOCK_BYTES32_VALUE);
+  }
+
+  function testBytesTryGetSuccess() public {
+    assertTrue(!s_addressToBytesMap.contains(address(this)));
+    assertTrue(s_addressToBytesMap.set(address(this), MOCK_BYTES_VALUE));
+    assertTrue(s_addressToBytesMap.contains(address(this)));
+    (bool success, bytes memory value) = s_addressToBytesMap.tryGet(address(this));
+    assertTrue(success);
+    assertEq(keccak256(value), keccak256(MOCK_BYTES_VALUE));
```
Comment (elatoskinas) [inline contracts/src/v0.8/shared/test/enumerable/EnumerableMapAddresses.t.sol]:
```suggestion
    assertEq(value, MOCK_BYTES_VALUE);
```

File: contracts/src/v0.8/shared/test/enumerable/EnumerableMapAddresses.t.sol [hunk 17]
```diff
@@ -140,37 +188,62 @@ contract EnumerableMapAddresses_tryGet is EnumerableMapAddressesTest {
...
     assertTrue(s_addressToBytes32Map.contains(address(this)));
-    assertTrue(s_addressToBytes32Map.get(address(this)) == MOCK_BYTES32_VALUE);
+    assertEq(s_addressToBytes32Map.get(address(this)), MOCK_BYTES32_VALUE);
+  }
+
+  function testBytesGetSuccess() public {
+    assertTrue(!s_addressToBytesMap.contains(address(this)));
+    assertTrue(s_addressToBytesMap.set(address(this), MOCK_BYTES_VALUE));
+    assertTrue(s_addressToBytesMap.contains(address(this)));
+    assertEq(keccak256(s_addressToBytesMap.get(address(this))), keccak256(MOCK_BYTES_VALUE));
```
Comment (elatoskinas) [inline contracts/src/v0.8/shared/test/enumerable/EnumerableMapAddresses.t.sol]:
```suggestion
    assertEq(s_addressToBytesMap.get(address(this)), MOCK_BYTES_VALUE);
```

File: contracts/src/v0.8/shared/test/enumerable/EnumerableMapAddresses.t.sol [hunk 18]
```diff
@@ -140,37 +188,62 @@ contract EnumerableMapAddresses_tryGet is EnumerableMapAddressesTest {
...
     assertTrue(s_addressToAddressMap.contains(address(this)));
-    assertTrue(s_addressToAddressMap.get(address(this), "EnumerableMapAddresses: nonexistent key") == address(this));
+    assertEq(s_addressToAddressMap.get(address(this), "EnumerableMapAddresses: nonexistent key"), address(this));
+  }
+
+  function testBytesGetErrorMessageSuccess() public {
+    assertTrue(!s_addressToBytesMap.contains(address(this)));
+    assertTrue(s_addressToBytesMap.set(address(this), MOCK_BYTES_VALUE));
+    assertTrue(s_addressToBytesMap.contains(address(this)));
+    assertEq(keccak256(s_addressToBytesMap.get(address(this))), keccak256(MOCK_BYTES_VALUE));
```
Comment (elatoskinas) [inline contracts/src/v0.8/shared/test/enumerable/EnumerableMapAddresses.t.sol]:
```suggestion
    assertEq(s_addressToBytesMap.get(address(this)), MOCK_BYTES_VALUE);
```

**Review by elatoskinas — DISMISSED** [review]

File: contracts/gas-snapshots/shared.gas-snapshot [hunk 19]
```diff
@@ -51,21 +51,29 @@ CallWithExactGas__callWithExactGasSafeReturnData:test_NoContractReverts() (gas:
 CallWithExactGas__callWithExactGasSafeReturnData:test_NoGasForCallExactCheckReverts() (gas: 16139)
```
Comment (elatoskinas) [inline contracts/gas-snapshots/shared.gas-snapshot:51]: CI is failing - requires snapshot fix

https://github.com/smartcontractkit/chainlink/actions/runs/10005679482/job/27656867201?pr=13890

Comment (defistar) [reply contracts/gas-snapshots/shared.gas-snapshot:51]: fixed it now

**Review by elatoskinas — APPROVED** [review]

**Review by makramkd — APPROVED** [review]

File: contracts/src/v0.8/shared/enumerable/EnumerableMapAddresses.sol [hunk 20]
```diff
@@ -1,12 +1,15 @@
 // SPDX-License-Identifier: MIT
+/* solhint-disable-next-line chainlink-solidity/prefix-internal-functions-with-underscore */
 pragma solidity ^0.8.0;
 
 import {EnumerableMap} from "../../vendor/openzeppelin-solidity/v4.8.3/contracts/utils/structs/EnumerableMap.sol";
+import {EnumerableMapBytes32} from "./EnumerableMapBytes32.sol";
 
 // TODO: the lib can be replaced with OZ v5.1 post-upgrade, which has AddressToAddressMap and AddressToBytes32Map
```
Comment (RayXpub) [inline contracts/src/v0.8/shared/enumerable/EnumerableMapAddresses.sol:8]: nit: I think this is no longer true as OZ v5 does not support `AddressToBytesMap`
